### ISSUE TITLE
What are the things you want to know about a foodtruck?

### ISSUE THREAD

**Issue by brekk** [foodraylabs/linter#1]
These can be:
  * Availability
    - time of day
    - day of week
  * Name of truck
  * Location of truck
  * Menu
  * Gluten / Allergy / whatever-filter free?
  * Reviews?

**Comment by stein935** [comment]
* Availability
  * current location (if any)
  * time of day
  * day of week
  * expected (after aggregating data at locations, possibly)
* Truck
  * type of food
  * owner
* Menu
  * item name
  * ingredients
  * calories 
  * serving size
  * review
  * descriptions
  * categories (gluten / allergy / whatever-filter booleans)

**Comment by brekk** [comment]
So I think we should start with the things we can get from the pages themselves:

```json
{
    "meta": {
        "lastRequest": 1439900225554,
        "version": "0.0.1"
    },
    "detail": {
        "title": "Señor Sisig: Filipino Fusion Food Truck",
        "description": "Señor Sisig is a Filipino Fusion Food Truck located in the San Francisco Bay Area.",
        "email": "info@senorsisig.com"
    },
    "menu": {
        "sisigtaco": {
            "name": "Sisig Taco",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "6\" corn tortilla, onions, lettuce & our cilantro cream sauce"
        },
        "thesenorsisigburrito": {
            "name": "The \"Señor\" Sisig Burrito",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "With Adobo garlic rice, pinto beans, lettuce, pico de gallo & our cilantro cream sauce"
        },
        "californiasisigburrito": {
            "name": "California Sisig Burrito",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "With french fries inside, shredded cheese, sour cream, guacamole & pico de gallo"
        },
        "sisigfries": {
            "name": "Sisig Fries",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "Crispy shoestring French fries topped with nacho cheese, sour cream, guacamole, pico de gallo & picked jalepeños"
        },
        "sisignachos": {
            "name": "Sisig Nachos",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "Corn tortilla chips topped with nacho cheese, sour cream, guacamole, pico de gallo & picked jalepeños"
        },
        "sisigwsteamedrice": {
            "name": "Sisig W/ Steamed Rice",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "Topped w/ diced onions, diced jalepeños & a side of achara"
        },
        "sisigsalad": {
            "name": "Sisig Salad",
            "option": {
                "protein": ["Pork", "Chicken", "Tofu"]
            },
            "description": "Shredded lettuce, pico de gallo, cilantro cream sauce & shredded cheese"
        },
        "tosilogburrito": {
            "name": "Tosilog Burrito",
            "availability": {
                "days": "(Only on Tuesday & Thursday)"
            },
            "description": "Filipino Sweet Pork, adobo garlic rice, fresh tomatoes and a fried egg. Served with a optional side of peppered vinegar. (Only on TUESDAY & THURSDAY)"
        }
    },
    "location": {
        "now": {
            // a reference to location.on.[Today's Date]
        },
        "on": {
            "2015-08-20": {
                "date": "2015-08-20",
                "title": "Off The Grid \"UN Plaza\"",
                "address": "Civic Center/UN Plaza, San Francisco, CA 94102",
                "open": true,
                "availability": {
                    "start": "2015-08-20 11:00:00",
                    "end": "2015-08-20 14:00:00"
                }
            }
        }
    }
}
```

I'll do the same to a couple more food truck sites, and then I'll write a thing for structuring how to do it, but I think we wanna start with parseable data that we can slice apart. Unfortunately most of these sites aren't written very well or structured consistently (the misspelled jalapeños above bugs me), but I think we can start to cobble together a common structure.

@stein935 I think the detailed info about the menu items makes sense, but we might need to parse it from a review site or somewhere with extended information.

I'm imagining that the `menu.sisigtaco.option.protein` and the `menu.tosilogburrito.availability.days` data structures are the pieces we'll need to standardize, 'cause I can see my way clear to scraping that info out, but structurally we'll need to be consistent.

Additionally, we'll probably have to throw some meta-data like `lastRequest` and `version` in there, 'cause if the page changes (broadly), we'll need to modify the scraper.

**Comment by ryanwaters** [comment]
Despite the misspellings this is much more data than I thought we would get
from a scrape, cuz like you said, those sites are generally pretty poor.

**Comment by brekk** [comment]
Yeah, this needs a second pass to convert the options and generally robustify things, but this will generate almost everything:

```coffeescript
xxx = (require 'x-ray')()

data = xxx 'http://senorsisig.com', {
    meta: {
        title: 'meta[property="og:title"]@content'
        description: 'meta[property="og:description"]@content'
        email: 'meta[property="og:email"]@content'
    }
    menu: xxx '#menu .row', [{
        name: '.menu-item h3'
        description: '.menu-item p ~ p'
        optionProten: '.menu-item em'
        picture: '.span8 img@src'
    }]
    location: 
        on: xxx '#loc-wrap section', [{
            date: '@data-wcal-date'
            title: '.location h5'
            address: '.location p a'
            availability: {
                start: '.time span:nth-child(1)'
                end: '.time span:nth-child(2)'
            }
        }]
}
data (error, object)->
    if error?
        console.log error
        throw error
    console.log JSON.stringify object, null, 4
```

**Comment by brekk** [comment]
Slightly cleaner:

```coffeescript
xxx = (require 'x-ray')()
_ = require 'lodash'
slug = require 'slug'
flatslug = (x)->
    return slug(x.toLowerCase()).replace(/-/g, '')

data = xxx 'http://senorsisig.com', {
    meta: {
        title: 'meta[property="og:title"]@content'
        description: 'meta[property="og:description"]@content'
        email: 'meta[property="og:email"]@content'
    }
    menu: xxx '#menu .row', [{
        name: '.menu-item h3'
        description: '.menu-item p ~ p'
        optionProtein: '.menu-item em'
        picture: '.span8 img@src'
    }]
    location: 
        on: xxx '#loc-wrap section', [{
            date: '@data-wcal-date'
            title: '.location h5'
            address: '.location p a'
            availability: {
                start: '.time span:nth-child(1)'
                end: '.time span:nth-child(2)'
            }
        }]
}
data (error, object)->
    if error?
        console.log error
        throw error
    object.menu = _(object.menu).map((item)->
        obj = {}
        if item.optionProtein?
            item.options = {
                protein: null
            }
        if item.optionProtein?
            prop = item.optionProtein.toLowerCase()
            reprop = _(prop.substr(1, prop.length-2).split(',')).map((x)-> return x.split(' ')).flatten().filter((x)-> return x isnt '').without('or', 'choice', 'of').value()
            item.options.protein = reprop
            delete item.optionProtein
        if item.description?
            item.description = item.description.trim()
        if item.name?
            item.name = item.name.trim()
        obj[flatslug(item.name)] = item
        return obj
    ).reduce (y, z)->
        return _.extend y, z
    , {}
    console.log JSON.stringify object, null, 4
```

**Comment by brekk** [comment]
With comments:

```coffeescript
"use strict"

xxx = (require 'x-ray')()
_ = require 'lodash'
slug = require 'slug'
flatslug = (x)->
    return slug(x.toLowerCase()).replace(/-/g, '')
j4 = (x)->
    return JSON.stringify(x, null, 4)
swappable = require './swappable'

# template for scraping
data = xxx 'http://senorsisig.com', {
    # descriptor data
    meta: {
        title: 'meta[property="og:title"]@content'
        description: 'meta[property="og:description"]@content'
        email: 'meta[property="og:email"]@content'
    }
    # the menu
    menu: xxx '#menu .row', [{
        name: '.menu-item h3'
        description: '.menu-item p ~ p'
        optionProtein: '.menu-item em'
        picture: '.span8 img@data-src'
    }]
    # the location
    location: 
        on: xxx '#loc-wrap section', [{
            date: '@data-wcal-date'
            title: '.location h5'
            address: '.location p a'
            availability: {
                start: '.time span:nth-child(1)'
                end: '.time span:nth-child(2)'
            }
        }]
}
popParens = (x)->
    if _.isString x
        return x.substr(1, x.length-2)
# call our returned function with a callback
data (error, object)->
    if error?
        console.log error
        throw error
    # restructure the menu
    object.menu = _(object.menu).map((item)->
        menuItem = {}
        item = swappable item
        # we're using the swappables
        if item.optionProtein?
            # so that we can do things like this:
            item.swap('optionProtein', 'options.protein')
        # and make assertions like this:
        if item.has('options.protein')
            tein = item.get('options.protein')
            # easier to clean up in lowercase (or any converted case)
            prop = tein.toLowerCase()
            # split on commas, then split each sub string by spaces
            # then flatten the arrays
            # then remove any other words
            reprop = _(popParens(prop).split(',')).map (x)-> return x.split(' ')
                                                  .flatten()
                                                  .without '', 'or', 'choice', 'of'
                                                  .value()
            # overwrite the original value
            item.set('options.protein', reprop)
        # cruft in the description
        if item.description?
            item.description = item.description.trim()
        # and the name
        if item.name?
            item.name = item.name.trim()
        # flatslug the name so we don't have any collisions
        menuItem[flatslug(item.name)] = item.valueOf()
        return menuItem
    ).reduce (collection, iter)->
        return _.extend collection, iter
    , {}
    # restructure the array of locations so that they are keyed by date
    object.location.on = _(object.location.on).map((item)->
        raw = {}
        raw[item.date] = item
        return raw
    ).reduce (collection, iter)->
        return _.extend collection, iter
    , {}
    # un-helpful data that needs to be eschewed
    if object.location.on.error?
        delete object.location.on.error
    console.log JSON.stringify object, null, 4
    return object
```
The `swappable` file, required above:
```coffeescript
"use strict"

Obus = require 'obus'
_ = require 'lodash'

popNested = (input)->
    return input.substr(0, input.lastIndexOf('.'))

Obus::swap = (original, changed, deleteAfterChange=true, deleteFully=false)->
    value = @get original
    unless value?
        return false
    @set changed, value
    if deleteAfterChange
        parts = original.split('.')
        if (parts.length > 1) and deleteFully
            return _.map parts, (part, index)->
                carried = null
                out = _.map [0..index], ()->
                    unless carried?
                        carried = original
                    else
                        carried = popNested carried
                    return carried
                return _.map out, (item)->
                    return @del item
        return @del original
    return @

Obus::give = (request)->
    value = @get request
    unless value?
        return false
    if _.isObject value
        $child = MakeSwappable value
        return $child
    return value

MakeSwappable = (sourceObject)->
    $o = new Obus()
    _.each sourceObject, (property, key)->
        $o.set key, property
    return $o

module.exports = MakeSwappable
```

**Comment by brekk** [comment]
And then here's an attempt at getting the Bacon Bacon truck's page scraped:

```coffeescript
"use strict"

xxx = (require 'x-ray')()
_ = require 'lodash'
slug = require 'slug'
flatslug = (x)->
    return slug(x.toLowerCase()).replace(/-/g, '')
j4 = (x)->
    return JSON.stringify(x, null, 4)
swappable = require './swappable'

menuDrillDown = '.wc-shortcodes-accordion-trigger' #'.wc-shortcodes-item' # > .wc-shortcodes-accordion-trigger'

# template for scraping
data = xxx 'http://baconbaconsf.com/the-truck', {
    # descriptor data
    meta: {
        title: 'title'
        description: 'meta[name="description"]@content'
        twitter: '.twitter-follow-button@href'
    }
    # the menu
    menuLeft: xxx '.wc-shortcodes-clearfix > .wc-shortcodes-column-first', [{
        name: 'h4'
        description: 'h4 ~ p'
    }]
    menuRight: xxx '.wc-shortcodes-clearfix > .wc-shortcodes-column-last', [{
        name: 'h4'
        description: 'h4 ~ p'
    }]
    # the location
    location: 
        on: xxx ".wc-shortcodes-accordion", [{
            date: ".wc-shortcodes-accordion-trigger b"
            title: ".wc-shortcodes-accordion-content b"
            address: ".wc-shortcodes-accordion-content a@href"
            availability: ".wc-shortcodes-accordion-content"
        }]
}
# call our returned function with a callback
data (error, object)->
    if error
        console.log error
        throw error
        return
    # join together the two lists that their dumb markup broke
    if object.menuLeft? and object.menuRight?
        object.menu = object.menuLeft.concat object.menuRight
        delete object.menuLeft
        delete object.menuRight
    # if we succeeded above
    if object.menu?
        # map the object so the keys are the flatslugged names
        object.menu = _(object.menu).map((item)->
            wrap = {}
            wrap[flatslug(item.name.toLowerCase())] = item
            return wrap
        ).reduce (coll, iter)->
            return _.extend coll, iter
        , {}
    if object?.meta?.twitter?
        object.meta.twitter = object.meta.twitter.split('.com/')[1]
    console.log "J4", j4 object, null, 4
    return object

```

**Comment by brekk** [comment]
Hmmm. The bacon bacon page is very hard to parse logically, 'cause there's bad markup on the page.

(You can see how far I've had to go to try and get structure in the `location.onX`s. I still don't think it'll work fully, as that `:nth-last-of-type` doesn't seem to work.)

```coffeescript
"use strict"

xxx = (require 'x-ray')()
_ = require 'lodash'
slug = require 'slug'
flatslug = (x)->
    return slug(x.toLowerCase()).replace(/-/g, '')
j4 = (x)->
    return JSON.stringify(x, null, 4)
swappable = require './swappable'

menuDrillDown = '.wc-shortcodes-accordion-trigger' #'.wc-shortcodes-item' # > .wc-shortcodes-accordion-trigger'

# template for scraping
data = xxx 'http://baconbaconsf.com/the-truck', {
    # descriptor data
    meta: {
        title: 'title'
        description: 'meta[name="description"]@content'
        twitter: '.twitter-follow-button@href'
    }
    # the menu
    menuLeft: xxx '.wc-shortcodes-clearfix > .wc-shortcodes-column-first', [{
        name: 'h4'
        description: 'h4 ~ p'
    }]
    menuRight: xxx '.wc-shortcodes-clearfix > .wc-shortcodes-column-last', [{
        name: 'h4'
        description: 'h4 ~ p'
    }]
    # the location
    location:
        on: xxx '.wc-shortcodes-accordion-content', [{
            titles: ['b']
            locations: ['a@href']
        }]
        on2: xxx '.wc-shortcodes-accordion-trigger + .wc-shortcodes-accordion-content', [{
            lastLink: 'a:nth-last-of-type(1)@href'
        }]
        on3: xxx '.wc-shortcodes-accordion-trigger', [{
            date: 'b'
        }]
        on4: xxx '.wc-shortcodes-accordion-content', ['div']
        # on5: ['.wc-shortcodes-clearfix:nth-of-type(1)@html']
            
}
# call our returned function with a callback
data (error, object)->
    if error
        console.log error
        throw error
        return
    # join together the two lists that their dumb markup broke
    if object.menuLeft? and object.menuRight?
        object.menu = object.menuLeft.concat object.menuRight
        delete object.menuLeft
        delete object.menuRight
    # if we succeeded above
    if object.menu?
        # map the object so the keys are the flatslugged names
        object.menu = _(object.menu).map((item)->
            wrap = {}
            wrap[flatslug(item.name.toLowerCase())] = item
            return wrap
        ).reduce (coll, iter)->
            return _.extend coll, iter
        , {}
    if object?.meta?.twitter?
        object.meta.twitter = object.meta.twitter.split('.com/')[1]
    console.log "J4", j4 object, null, 4
    return object
```

Anyways, we may have to come up with a contingency approach for cases like these, but for now I'll try and treat this as an edge case.
